In [1]:
import pandas as pd

In [6]:
df = pd.read_excel("C:/Users/IBRA/Desktop/online-retail-sales-analysis/data/raw/Online Retail.xlsx")

In [7]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [8]:
df.shape

(541909, 8)

In [9]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [11]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [12]:
df.duplicated().sum()

np.int64(5268)

In [13]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


## Initial Data Quality Observations

The dataset contains 541,909 rows and 8 columns.

The initial data inspection shows:

- `Description` has 1,454 missing values.
- `CustomerID` has 135,080 missing values.
- All other columns have no missing values.
- `InvoiceDate` is already stored as a datetime column.
- `Quantity` is stored as an integer.
- `UnitPrice` is stored as a decimal number.
- `CustomerID` is stored as a float, but it should later be treated as an identifier rather than a numeric value.

The missing values in `CustomerID` are important because customer-level analysis requires customer identifiers. However, these rows may still be useful for general sales and product analysis.

## Statistical Summary

In [15]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386048,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


## Statistical Summary Observations

The statistical summary shows several important data quality issues:

- The `Quantity` column contains negative values, with a minimum value of `-80,995`. This may indicate cancelled orders, returns, or data entry issues.
- The maximum `Quantity` value is `80,995`, which is unusually high and should be investigated later.
- The `UnitPrice` column contains a negative value, with a minimum value of `-11,062.06`. Since product prices should not normally be negative, this will need to be handled during data cleaning.
- The maximum `UnitPrice` value is `38,970`, which may represent an outlier or special transaction.
- The dataset covers transactions from `2010-12-01` to `2011-12-09`.
- `CustomerID` is shown in the statistical summary, but it should not be interpreted mathematically because it is an identifier, not a real numeric measurement.

These findings confirm that the dataset needs cleaning before analysis.

## Duplicate Rows

The dataset contains 5,268 duplicated rows.

Duplicate transactions may affect the accuracy of sales, product, and customer analysis, so they will need to be removed or handled during the data cleaning stage.

In [16]:
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

## Unique Values Observations

The dataset contains:

- 25,900 unique invoice numbers.
- 4,070 unique stock codes, representing different product/item codes.
- 4,223 unique product descriptions.
- 4,372 unique customers.
- 38 unique countries.

The number of unique product descriptions is slightly higher than the number of stock codes. This may mean that some products have inconsistent descriptions, repeated descriptions, or special non-product transaction entries. This should be investigated later during data cleaning.

The dataset appears suitable for several types of analysis, including invoice-level analysis, product analysis, customer analysis, country analysis, and market basket analysis.

## Cancelled Invoices

In [19]:
df['InvoiceNo'].astype(str).str.startswith('C').sum()

np.int64(9288)

## Cancelled Invoice Observations

The dataset contains 9,288 rows where the invoice number starts with `C`.

In this dataset, invoice numbers beginning with `C` represent cancelled transactions. These rows are important because they may explain some of the negative quantity values found in the statistical summary.

Cancelled invoices should not be removed yet in the data understanding stage. They will be handled later during data cleaning, depending on whether the analysis focuses on completed sales only or includes cancellations/returns.

In [20]:
(df['Quantity'] < 0).sum()

np.int64(10624)

## Negative Quantity Observations

The dataset contains 10,624 rows where `Quantity` is less than zero.

Negative quantities may represent cancelled transactions, returned items, or correction entries. Since cancelled invoices were already identified using invoice numbers beginning with `C`, the relationship between cancelled invoices and negative quantities should be investigated during data cleaning.

These rows should not be removed yet. They will be handled in the data cleaning stage after understanding whether they represent cancellations, returns, or invalid transactions.

In [21]:
(df['UnitPrice'] <= 0).sum()

np.int64(2517)

## Zero or Negative Unit Price Observations

The dataset contains 2,517 rows where `UnitPrice` is less than or equal to zero.

A zero or negative unit price is not normal for completed sales transactions because products should usually have a positive selling price. These rows may represent cancelled transactions, adjustments, missing prices, free items, or data entry issues.

These rows should not be removed yet. They will be handled later during the data cleaning stage after checking their relationship with cancellations and other unusual transaction records.

In [22]:
df['InvoiceDate'].min(), df['InvoiceDate'].max()

(Timestamp('2010-12-01 08:26:00'), Timestamp('2011-12-09 12:50:00'))

## Date Range Observations

The dataset covers transactions from `2010-12-01` to `2011-12-09`.

This means the dataset contains slightly more than one year of transaction data. Because the data covers a full year plus several extra days, it can be used to analyze monthly sales trends, seasonal patterns, and changes in customer purchasing behavior over time.

In [23]:
df['Country'].nunique()

38

## Country Coverage Observations

The dataset contains transactions from 38 different countries.

This makes the dataset useful for country-level analysis, such as identifying the countries with the highest number of transactions, strongest sales contribution, and potential geographic concentration of customers.

In [24]:
df['Country'].value_counts().head(10)

Country
United Kingdom    495478
Germany             9495
France              8557
EIRE                8196
Spain               2533
Netherlands         2371
Belgium             2069
Switzerland         2002
Portugal            1519
Australia           1259
Name: count, dtype: int64

## Top Countries Observations

The top 10 countries by number of transaction rows are:

1. United Kingdom: 495,478 rows
2. Germany: 9,495 rows
3. France: 8,557 rows
4. EIRE: 8,196 rows
5. Spain: 2,533 rows
6. Netherlands: 2,371 rows
7. Belgium: 2,069 rows
8. Switzerland: 2,002 rows
9. Portugal: 1,519 rows
10. Australia: 1,259 rows

The United Kingdom strongly dominates the dataset, representing the majority of transaction records. This is expected because the dataset comes from a UK-based online retail business.

Because of this imbalance, country-level analysis should consider that results may be heavily influenced by the United Kingdom. Later analysis may compare the United Kingdom separately from international customers.

# Data Understanding Summary

The data understanding stage identified several important points:

- The dataset contains 541,909 rows and 8 columns.
- There are 5,268 duplicated rows.
- `Description` has 1,454 missing values.
- `CustomerID` has 135,080 missing values.
- There are 25,900 unique invoices, 4,070 unique stock codes, 4,372 unique customers, and 38 countries.
- There are 9,288 cancelled invoice rows where `InvoiceNo` starts with `C`.
- There are 10,624 rows with negative quantities.
- There are 2,517 rows with zero or negative unit prices.
- The dataset covers transactions from `2010-12-01` to `2011-12-09`.
- The United Kingdom is the dominant country in the dataset.

These findings show that the dataset is suitable for sales, customer, product, country, and market basket analysis. However, the data must be cleaned before analysis because it contains missing values, duplicates, cancellations, negative quantities, and invalid prices.